appsec-kit-v8/src/main/java/com/vaadin/appsec/v8/service/NotificationInitListener.java

```diff
@@ -33,80 +37,51 @@ public class NotificationInitListener extends AbstractInitListener {
     private static final Logger LOGGER = LoggerFactory
             .getLogger(NotificationInitListener.class);
 
-    private static final String NOTIFIED_UIS_SESSION_PARAM = "vaadin-appsec-kit-notified-uis";
-    /**
-     * New UI instance check interval (in ms). Determines how often the session
-     * is checked for new UI instances which might need a notification shown.
-     */
-    private static final int NOTIFICATION_CHECK_INTERVAL = 5000;
-
     /**
      * Notification timeout (in ms). Set to 24 hours to essentially make it
      * persistent until either the appsec link is clicked or the notification is
      * dismissed.
      */
     private static final int NOTIFICATION_DELAY = 24 * 60 * 60 * 1000;
 
+    private final Map<VaadinSession, Registration> scanEventRegistrations = new ConcurrentHashMap<>();
+
     @Override
     public void serviceInit(ServiceInitEvent event) {
-        VaadinService vaadinService = event.getSource();
-        if (isDebugMode(vaadinService)) {
-            vaadinService.addSessionInitListener(e -> {
-                e.getSession().addUIProvider(new AppSecUIProvider());
-                createNotificationThread(e.getSession()).start();
-            });
-            LOGGER.info("NotificationInitListener initialized.");
+        VaadinService service = event.getSource();
+        if (isDebugMode(service)) {
+            service.addSessionInitListener(this::subscribeSessionToScanEvents);
+            service.addSessionDestroyListener(this::removeSessionRegistration);
+            LOGGER.info("Subscribed to AppSec Kit scan events");
         }
     }
 
-    private Thread createNotificationThread(final VaadinSession session) {
-        return new Thread(() -> {
-            try {
-                LOGGER.debug(
-                        "NotificationInitListener notification thread initialized.");
-
-                Thread.sleep(NOTIFICATION_CHECK_INTERVAL);
+    private void removeSessionRegistration(SessionDestroyEvent e) {
+        VaadinSession session = e.getSession();
+        Registration registration = scanEventRegistrations.get(session);
+        if (registration != null) {
+            registration.remove();
+        }
+    }
 
-                while (isSessionOpen(session)) {
-                    session.access(() -> {
```

Comment:
The same applies for this as for the v7 version.

Reply:
Fixed also for V7.